rust/main/agents/relayer/src/msg/pending_message.rs

```diff
@@ -250,81 +253,23 @@ impl PendingOperation for PendingMessage {
             return PendingOperationResult::Drop;
         }
 
-        let ism_address = match self
-            .ctx
-            .destination_mailbox
-            .recipient_ism(self.message.recipient)
-            .await
-        {
-            Ok(ism_address) => ism_address,
-            Err(err) => {
-                return self.on_reprepare(Some(err), ReprepareReason::ErrorFetchingIsmAddress);
-            }
-        };
-
-        let message_metadata_builder = match MessageMetadataBuilder::new(
-            self.ctx.metadata_builder.clone(),
-            ism_address,
-            &self.message,
-        )
-        .await
-        {
-            Ok(message_metadata_builder) => message_metadata_builder,
-            Err(err) => {
-                return self.on_reprepare(Some(err), ReprepareReason::ErrorGettingMetadataBuilder);
-            }
-        };
-
-        let params = MessageMetadataBuildParams::default();
-
-        let metadata_bytes = match message_metadata_builder
-            .build(ism_address, &self.message, params)
-            .await
-        {
-            Ok(metadata) => {
-                let metadata_bytes = metadata.to_vec();
-                self.metadata = Some(metadata_bytes.clone());
-                metadata_bytes
+        let metadata_bytes = match self.metadata.as_ref() {
+            // if metadata is already built and is not too stale
+            Some((metadata, built_time))
+                if Instant::now().duration_since(*built_time).as_secs() < STALE_METADATA_SECS =>
+            {
+                tracing::debug!("Reusing cached metadata");
```

Comment:
would suggest either having this as a const or adding a comment that e2e relies on this

Reply:
Good point